GoogleContainerTools/skaffold
# feat: add toleration for GKE ARM nodes taint

Fixes: #7779 

**Description**

The GKE ARM nodes are tainted with `kubernetes.io/arch="arm64":NoSchedule` by default to prevent workloads from being scheduled. In this PR, Skaffold will automatically add the corresponding toleration to the manifest for ARM image Pods against GKE clusters. 
This can be turned off with the flag `--enable-gke-arm-node-toleration=false`.

**Testing Instruction**
* Use a GKE ARM node cluster (node pool version `1.23.7-gke.1400` or greater). Googlers can use [this](https://pantheon.corp.google.com/kubernetes/clusters/details/us-central1-a/integration-tests-hybird/details?project=k8s-skaffold) cluster
* In the `integration/examples/cross-platform-builds` project, with Skaffold built from `main` branch HEAD, run:
  ```
  skaffold dev --default-repo=gcr.io/k8s-skaffold --platform=linux/arm64
  ```
  This should fail due to the node Taint.
  ```
  Starting deploy...
   - pod/getting-started created
  Waiting for deployments to stabilize...
   - pods: Unschedulable: 0/1 nodes available: 
      - pod/getting-started: Unschedulable: 0/1 nodes available: 
   - pods: 0/3 nodes are available: 1 node(s) didn't match Pod's node affinity/selector, 2 node(s) had taint {kubernetes.io/arch: arm64}, that the pod didn't tolerate.
      - pod/getting-started: 0/3 nodes are available: 1 node(s) didn't match Pod's node affinity/selector, 2 node(s) had taint {kubernetes.io/arch: arm64}, that the pod didn't tolerate.
   - pods failed. Error: 0/3 nodes are available: 1 node(s) didn't match Pod's node affinity/selector, 2 node(s) had taint {kubernetes.io/arch: arm64}, that the pod didn't tolerate..
  ```
* Now, build Skaffold from this PR branch and run:
  ```
  skaffold dev --default-repo=gcr.io/k8s-skaffold --platform=linux/arm64
  ```
  This should now succeed.
  ```
  Starting deploy...
   - pod/getting-started created
  Waiting for deployments to stabilize...
   - pods is ready.
  Deployments stabilized in 6.941 seconds
  Press Ctrl+C to exit
  Watching for changes...
  [getting-started] Hello world! Running on linux/arm64
  ```
